src/live-programming.rst

```diff
@@ -20,9 +20,203 @@ Data Analysis Environments
 Debugging
 =========
 
+  "The EDSAC was on the top floor of the building and the tape-punching
+  and editing equipment one floor below... It was on one of my journeys
+  between the EDSAC room and the punching equipment that...the realization
+  came over me with full force that a good part of the remainder of my
+  life was going to be spent in finding errors in my own programs."
+
+  -- *Memoirs of a Computer Pioneer*, Maurice Wilkes
+
+Much of the activity of programming involves **debugging**, i.e.,
+diagnosing and preventing undesired program behavior.
+Wilkes's early realization of this fact was echoed and amplified in a 2002
+study of the U.S. software industry, which found that software engineers
+spent an average of 70-80% of their time testing and debugging, with
+the average bug taking 17.4 hours to fix.
+Thus, debuggers form a crucial component of an effective programming system.
+(needs better finish)
+
+Software bugs manifest at three levels. Typically a bug is first
+observed as a **runtime failure**, an instance of external program
+behavior that does not comply with the intended design. Underlying
+the runtime failure are **runtime faults**, aberrations
+in machine or program state (e.g., a wrong value in a CPU register,
+an uninitialized object) that lead to the noncompliant behavior.
+Finally, causing the runtime faults are the **software errors** in
+the source code.
+
+Using these terms, debugging may be understood
+more precisely as the collective processes of determining what runtime
+faults led to a runtime failure, determining what software errors
+led to those faults, and modifying the code to prevent the faults from
```

Comment:
"modifying the code to prevent the faults from occurring", i.e. eliminating the errors?